### ISSUE TITLE
Large error between distance travelled along center line and distance actually travelled

### ISSUE THREAD

**Issue by AngelosSE** [nutonomy/nuscenes-devkit#883]
If the initial and final positions of a road user is projected on the closest centerline and the road user's trajectory of positions between initial and final positions is close to the centerline, then I expect that computing the arclength between the projected posititions should give approximately the same distance as integrating the velocity of the road user as it travels from its initial position to its final position. However, there exists cases in the data where this is not the case as shown in the code attached below. Is the source of such errors known? In one of the attached examples I get an error of 4.26 m!

```
from nuscenes.map_expansion.map_api import NuScenesMap
from nuscenes.map_expansion import arcline_path_utils
from nuscenes.prediction import PredictHelper

from nuscenes import NuScenes
import nuscenes.map_expansion.arcline_path_utils as ARC
from nuscenes.eval.prediction.splits import get_prediction_challenge_split

def get_map_name(sample_token,nusc):
    sample_table = nusc.get('sample',sample_token)
    scene_token = sample_table['scene_token']
    log_token = nusc.get('scene',scene_token)['log_token']
    return nusc.get('log',log_token)['location']
    
def init(idx):
    DATAROOT = '/datasets/nuscenes/v1.0-mini/'
    nusc = NuScenes('v1.0-mini', DATAROOT,verbose=False)    
    data = get_prediction_challenge_split("mini_val", dataroot=DATAROOT)
    helper = PredictHelper(nusc)
    instance_token, sample_token = data[idx].split("_")
    map_name = get_map_name(sample_token,nusc)
    nusc_map = NuScenesMap(DATAROOT, map_name)
    return helper,nusc_map,instance_token,sample_token

def determine_maximum_normal_distance(lane,positions,resolution_meters):
    maximum_normal_distance = 0
    for position in positions:
        pose_on_lane,_ = ARC.project_pose_to_lane(position ,lane ,resolution_meters)
        position_on_lane = np.array(pose_on_lane[:2])
        normal_distance = np.linalg.norm(position-position_on_lane)
        if normal_distance > maximum_normal_distance:
            maximum_normal_distance = normal_distance
    return maximum_normal_distance

def main(idx,xlim,ylim):
    helper,nusc_map,instance_token,sample_token = init(idx)
    current_pose = helper.get_sample_annotation(instance_token,sample_token)['translation']
    current_position = np.array(current_pose[:2])    
    closest_lane = nusc_map.get_closest_lane(*current_position, radius=1)
    lane = nusc_map.get_arcline_path(closest_lane)
    resolution_meters = 0.01
    lane_poses = arcline_path_utils.discretize_lane(lane, resolution_meters)
    horizon = 6
    positions = helper.get_future_for_agent(instance_token
                                                ,sample_token
                                                ,seconds=horizon
                                                ,in_agent_frame=False)
    current_pose_on_lane,current_arclength = ARC.project_pose_to_lane(current_pose
                                                                ,lane
                                                                ,resolution_meters)
    current_position_on_lane = np.array(current_pose_on_lane[:2])
    future_sample_annotations_global = helper.get_future_for_agent(instance_token
                                                                ,sample_token
                                                                ,seconds=horizon
                                                                ,in_agent_frame=False
                                                                ,just_xy=False)
    final_sample_annotation_global = future_sample_annotations_global[-1]
    final_pose = final_sample_annotation_global['translation']
    final_pose_on_lane,final_arclength = ARC.project_pose_to_lane(final_pose
                                                            ,lane
                                                            ,resolution_meters)
    final_position_on_lane = np.array(final_pose_on_lane[:2])
    distance_on_lane = final_arclength-current_arclength
    distance_integrated = np.sum(np.linalg.norm(np.diff(positions,axis=0),axis=1))
    distance_error = np.abs(distance_on_lane-distance_integrated)
    tmp = np.row_stack([current_position,positions])
    max_normal_dist = determine_maximum_normal_distance(lane,tmp,resolution_meters)

    # plotting
    _,ax = plt.subplots()
    ax.set_aspect('equal')
    ax.set_xlim(xlim)
    ax.set_ylim(ylim)
    ax.plot([x for x,*_ in lane_poses],[y for _,y,_ in lane_poses],'.',label='')
    ax.plot(*current_position_on_lane,'.',color='orange',label='inital position on lane')
    ax.plot(*final_position_on_lane,'x',color='orange', label ='final position on lane')
    ax.plot(*positions.T,'r.', label='positions')
    ax.plot(*current_position,'or',mfc='none',label = 'initial position')
    ax.legend()

    # printing
    #print(f'{current_arclength=}')
    #print(f'{current_pose=}')
    #print(f'{final_arclength=}')
    #print(f'{final_pose=}')
    #print(f'{positions.shape=}')
    print(f'{distance_on_lane=:.2f}')
    print(f'{distance_integrated=:.2f}')
    print(f'{distance_error=:.2f}')
    print(f'{max_normal_dist=:.2f}')


print('\n')
print('Experiment showing large error:')
main(12,(580,625),(1630,1660))
print('\n')
print('Experiment showing small error:')
main(35,(662,666),(1601,1607))
plt.show()
```

**Comment by AngelosSE** [comment]
The provided arclength parameterization seems correct. When running the folowing code I get the expected result:

```
def plot_curve_parameter_VS_distance_travelled(lane):
    """
    If the parameterization is the arclength parameterization, then the result
    is the line y=x.
    """
    pose = arcline_path_utils.pose_at_length(lane[0],1e-6)
    position = np.array(pose[:2])
    ts = np.linspace(1e-6,166,1000)
    distances = []
    for t in ts:
        next_pose = arcline_path_utils.pose_at_length(lane[0],t)
        next_position = np.array(next_pose[:2])
        distances.append(np.linalg.norm(next_position-position))
        position = next_position
    cum_distances = np.cumsum(distances)
    _,ax = plt.subplots()
    ax.set_aspect('equal')
    ax.plot(ts,cum_distances)
    print(ts[-5:])
    print(cum_distances[-5:])
```

**Comment by AngelosSE** [comment]
The method `arcline_path_utils.project_pose_to_lane` seems correct, I get the expected result when running:

```
def check_project_pose_to_lane(lane,current_pose,final_pose):
    resolution_meters=0.01
    my_current_arclength = find_arclength_of_normal_proj(current_pose[:2],lane)
    _,current_arclength = arcline_path_utils.project_pose_to_lane(current_pose
                                                            ,lane
                                                            ,resolution_meters)
    _,final_arclength = arcline_path_utils.project_pose_to_lane(final_pose
                                                        ,lane
                                                        ,resolution_meters)
    my_final_arclength = find_arclength_of_normal_proj(final_pose[:2],lane)
    print(np.abs(current_arclength-my_current_arclength))
    print(np.abs(final_arclength-my_final_arclength))

def find_arclength_of_normal_proj(position,lane):
    N = 10000
    positions = np.full((N,2),np.nan)
    ts = np.linspace(1e-6,166,N)
    for i, t in enumerate(ts):
        positions[i] = arcline_path_utils.pose_at_length(lane[0],t)[:2]
    idx = np.argmin(np.linalg.norm(positions - position,axis=1))
    return ts[idx]
```

**Comment by AngelosSE** [comment]
The error was on my end. When computing the initial arclength I used the sample at the current timestep, but when integrated the speed I used timestep+1. The current speed of the vehicle was 8.77 and with a sampling time of 0.5 seconds this gives an error 4.39 m which matches well the error I reported. After the fix I get an error of 0.11 m which is reasonable.
